### PR TITLE
More terminal classification

### PR DESCRIPTION
**Pull request by hamishknight**
Add terminal classification for anchors, special escape characters, and meta characters, parameterized by whether or not we're in a custom character class.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

File: Sources/Regex/Anchor.swift [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/// An token that matches against a certain position in the subject.
+public enum Anchor: String, Hashable {
+  case lineStart = "^"
+  case lineEnd = "$"
+  case wordBoundary = "\\b"
+  case nonWordBoundary = "\\B"
```
Comment (milseman) [inline Sources/Regex/Anchor.swift:6]: I'm fine with more broadly calling built-in assertions "anchors", which is what `.NET` calls the IIRC

File: Sources/Regex/Parse/Token.swift [hunk 2]
```diff
@@ -19,6 +19,9 @@ enum Token: Hashable {
   case setOperator(SetOperator)
   case character(Character, isEscaped: Bool)
   case unicodeScalar(UnicodeScalar)
+  case builtinCharClass(CharacterClass)
+  case anchor(Anchor)
+  case specialCharEscape(SpecialCharacterEscape)
```
Comment (milseman) [inline Sources/Regex/Parse/Token.swift:24]: Have you thought about what the AST representation should be?

File: Sources/Regex/Parse/Token.swift [hunk 3]
```diff
@@ -41,16 +44,23 @@ extension Token {
...
   }
 
+  enum SpecialCharacterEscape: String, Hashable {
+    case tab = "\\t"
+    case carriageReturn = "\\r"
+    case backspace = "\\b"
+    case formFeed = "\\f"
+    case bell = "\\a"
+    case escape = "\\e"
+  }
```
Comment (milseman) [inline Sources/Regex/Parse/Token.swift:62]: Why not `Character`, which would synthesize the switch for you?

File: Sources/Regex/Parse/Token.swift [hunk 4]
```diff
@@ -79,13 +89,19 @@ extension Token.MetaCharacter: CustomStringConvertible {
...
 extension Token: CustomStringConvertible {
   var description: String {
     switch self {
     case .meta(let meta): return meta.description
     case .setOperator(let op): return op.description
     case .character(let c, _): return c.halfWidthCornerQuoted
     case .unicodeScalar(let u): return "U\(u.halfWidthCornerQuoted)"
+    case .builtinCharClass(let cc): return cc.description
+    case .anchor(let anchor): return anchor.description
+    case .specialCharEscape(let special): return special.description
```
Comment (milseman) [inline Sources/Regex/Parse/Token.swift:104]: What is "special" about this?

File: Sources/Regex/Parse/TokenClassification.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    default:
+      // By default, no other metacharacters exist in custom character
+      // classes.
+      if inCustomCharClass { return nil }
+    }
+    return mc
+  }
+
+  /// Classify a given terminal character
+  static func classifyTerminal(
```
Comment (milseman) [inline Sources/Regex/Parse/TokenClassification.swift:87]: So this is the intramodule programming interface. Could you provide a lexical grammar as a doc comment to show what's going on?

File: Tests/RegexTests/ParseTests.swift [hunk 6]
```diff
@@ -111,6 +117,23 @@ extension RegexTests {
       .leftSquareBracket, .rightSquareBracket,
       .setOperator(.doubleAmpersand), .rightSquareBracket,
       .rightSquareBracket, "&", "&")
+
+    lexTest("$\\A\\B[\\A\\B$]", .anchor(.lineEnd), .anchor(.stringStart),
```
Comment (milseman) [inline Tests/RegexTests/ParseTests.swift:121]: You can use raw string literals :-)

I didn't have them back in 2017.

File: Tests/RegexTests/ParseTests.swift [hunk 7]
```diff
@@ -111,6 +117,23 @@ extension RegexTests {
...
+
+    lexTest("\\t\\r\\f\\a\\e[\\t\\r\\f\\a\\e]",
+            specialChars + [.leftSquareBracket] + specialChars +
+            [.rightSquareBracket])
+
+    // \b is a word boundary outside of a character class, otherwise it's
+    // backspace.
+    lexTest("[\\b]\\b", .leftSquareBracket, .specialCharEscape(.backspace),
+            .rightSquareBracket, .anchor(.wordBoundary))
+    lexTest("[\\b", .leftSquareBracket, .specialCharEscape(.backspace))
```
Comment (milseman) [inline Tests/RegexTests/ParseTests.swift:136]: Is there a test of the 3 meanings of `^`?

**Review by milseman — COMMENTED** [review]
Overall I'm excited to see terminal classification. Just needs some comments, etc., either way thanks for writing this in a way that's explicitly parameterized so that it's easier to adopt to a new lexing design.

File: Tests/RegexTests/ParseTests.swift (hunk 7 above)
Comment (hamishknight) [reply Tests/RegexTests/ParseTests.swift:136]: Not in the same test case, but there are tests that exercise the different meanings

File: Sources/Regex/Parse/Token.swift (hunk 4 above)
Comment (hamishknight) [reply Sources/Regex/Parse/Token.swift:104]: It's just what I decided to call it, I don't suppose you have any suggestions on an alternative name?

File: Sources/Regex/Parse/Token.swift (hunk 3 above)
Comment (hamishknight) [reply Sources/Regex/Parse/Token.swift:62]: It felt weird to just have the raw values be e.g `t` without the `\`, and there's not many of them so it didn't really feel like much of a chore to spell out. I'm happy to change it tho.

File: Sources/Regex/Parse/Token.swift (hunk 2 above)
Comment (hamishknight) [reply Sources/Regex/Parse/Token.swift:24]: Yeah I'm currently not too sure, it seems they could either be their own category or just be folded into character or unicode scalar. I don't really feel too strongly either way, I don't suppose you have any thoughts?

File: Sources/Regex/Anchor.swift (hunk 1 above)
Comment (hamishknight) [reply Sources/Regex/Anchor.swift:6]: Yeah I was going off what Oniguruma calls them

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Comment by milseman** [comment]
@hamishknight I want to take this and rebase my stuff on top. Feel free to open a new PR for the comments or small changes.

**Pull request merged**